crates/net/network/src/transactions/fetcher.rs

```diff
@@ -745,6 +744,14 @@ impl<N: NetworkPrimitives> TransactionFetcher<N> {
         // try to fill request by checking if any other hashes pending fetch (in lru order) are
         // also seen by peer
         for hash in self.hashes_pending_fetch.iter() {
+            // Early exit if there's no budget
+            if let Some(ref mut bud) = budget_fill_request {
+                *bud = bud.saturating_sub(1);
+                if *bud == 0 {
+                    return
+                }
+            }
```

Comment:
same, q


this budget logic is super flawed, this shouldn't really need any of this.

I'm pretty sure we can get rid of this entirely by improving the way we track peer <-> hashes

Reply:
Should this be new PR, I'd need a bit more context but I think I can do it as well.

Reply:
this would require more significant changes, we can think about this separately

Reply:
Cool, reverted for now. Just a note it will affect perf.